sklearn/_fast_gradient_boosting/grower.py

```diff
@@ -0,0 +1,481 @@
+                smallest_child = left_child_node
+                largest_child = right_child_node
+            else:
+                smallest_child = right_child_node
+                largest_child = left_child_node
+
+            # We use the brute O(n_samples) method on the child that has the
+            # smallest number of samples, and the subtraction trick O(n_bins)
+            # on the other one.
+            tic = time()
+            smallest_child.histograms = \
+                self.histogram_builder.compute_histograms_brute(
+                    smallest_child.sample_indices)
+            largest_child.histograms = \
+                self.histogram_builder.compute_histograms_subtraction(
+                    node.histograms, smallest_child.histograms)
+            self.total_compute_hist_time += time() - tic
+
+            tic = time()
+            if should_split_left:
+                self._compute_best_split_and_push(left_child_node)
+            if should_split_right:
+                self._compute_best_split_and_push(right_child_node)
+            self.total_find_split_time += time() - tic
+
+        return left_child_node, right_child_node
+
+    def can_split_further(self):
+        """Return True if there are still nodes to split."""
+        return len(self.splittable_nodes) >= 1
+
+    def _finalize_leaf(self, node):
+        """Compute the prediction value that minimizes the objective function.
+
+        This sets the node.value attribute (node is a leaf iff node.value is
+        not None).
+
+        See Equation 5 of:
+        XGBoost: A Scalable Tree Boosting System, T. Chen, C. Guestrin, 2016
+        https://arxiv.org/abs/1603.02754
+        """
+        node.value = -self.shrinkage * node.sum_gradients / (
+            node.sum_hessians + self.splitter.l2_regularization)
+        self.finalized_leaves.append(node)
+
+    def _finalize_splittable_nodes(self):
+        """Transform all splittable nodes into leaves.
+
+        Used when some constraint is met e.g. maximum number of leaves or
+        maximum depth."""
+        while len(self.splittable_nodes) > 0:
+            node = self.splittable_nodes.pop()
+            self._finalize_leaf(node)
+
+    def make_predictor(self, bin_thresholds=None):
+        """Make a TreePredictor object out of the current tree.
+
+        Parameters
+        ----------
+        bin_thresholds : array-like of floats, optional (default=None)
+            The actual thresholds values of each bin.
+
+        Returns
+        -------
+        A TreePredictor object.
+        """
+        predictor_nodes = np.zeros(self.n_nodes, dtype=PREDICTOR_RECORD_DTYPE)
+        _fill_predictor_node_array(predictor_nodes, self.root,
+                                   bin_thresholds=bin_thresholds)
+        return TreePredictor(predictor_nodes)
+
+
+def _fill_predictor_node_array(predictor_nodes, grower_node,
+                               bin_thresholds=None, next_free_idx=0):
```

Comment:
It can be None in the tests where sometimes we only care about pre-binned data.

I'll remove the default for `bin_threholds` but if you don't mind I'll keep it for `next_free_idx=0` since this function should only really be called with `next_free_idx=0`, except on recursive calls. If we were really rigourous we would have a wrapper for this but I feel like it's overkill.

Reply:
I'd be fine with that.